src/NATS.Client.Core/NatsTlsOpts.cs

```diff
@@ -156,32 +145,181 @@ internal bool TryTls(NatsUri uri)
-    /// </summary>
-    private static X509Certificate2Collection LoadCertsFromMultiPem(ReadOnlySpan<char> certPem, ReadOnlySpan<char> keyPem)
+    internal async ValueTask<SslClientAuthenticationOptions> AuthenticateAsClientOptionsAsync(NatsUri uri)
     {
-        var multiPemCertificateCollection = new X509Certificate2Collection();
-        var addKey = true;
+#pragma warning disable CS0618 // Type or member is obsolete
+        if (EffectiveMode(uri) == TlsMode.Disable)
+        {
+            throw new InvalidOperationException("TLS is not permitted when TlsMode is set to Disable");
+        }
 
-        while (PemEncoding.TryFind(certPem, out var fields))
+        var options = new SslClientAuthenticationOptions
         {
-            X509Certificate2 certificate;
+            TargetHost = uri.Host,
 
-            if (addKey)
+            CertificateRevocationCheckMode = CertificateRevocationCheckMode,
+            EnabledSslProtocols = SslProtocols.Tls12 | SslProtocols.Tls13,
+        };
+
+        // validation
+        switch (this)
+        {
+        case { CertFile: not null, KeyFile: null } or { KeyFile: not null, CertFile: null }:
+            throw new ArgumentException("NatsTlsOpts.CertFile and NatsTlsOpts.KeyFile must both be set");
+        case { CertFile: not null, KeyFile: not null, LoadClientCert: not null }:
+            throw new ArgumentException("NatsTlsOpts.CertFile/KeyFile and NatsTlsOpts.LoadClientCert cannot both be set");
+        case { CaFile: not null, LoadCaCerts: not null }:
+            throw new ArgumentException("NatsTlsOpts.CaFile and NatsTlsOpts.LoadCaCerts cannot both be set");
+        }
+
+        if (CaFile != null)
+        {
+            options.LoadCaCertsFromPem(await File.ReadAllTextAsync(CaFile).ConfigureAwait(false));
+        }
+
+        if (LoadCaCerts != null)
+        {
+            options.LoadCaCertsFromX509(await LoadCaCerts().ConfigureAwait(false));
+        }
+
+        if (CertFile != null && KeyFile != null)
+        {
+            options.LoadClientCertFromPem(
+                await File.ReadAllTextAsync(CertFile).ConfigureAwait(false),
+                await File.ReadAllTextAsync(KeyFile).ConfigureAwait(false));
+        }
+
+        if (LoadClientCert != null)
+        {
+            options.LoadClientCertFromX509(await LoadClientCert().ConfigureAwait(false));
+        }
+
+        if (InsecureSkipVerify)
+        {
+            options.InsecureSkipVerify();
+        }
+
+        if (ConfigureClientAuthentication != null)
+        {
+            await ConfigureClientAuthentication(options).ConfigureAwait(false);
+        }
+
+        return options;
+#pragma warning restore CS0618 // Type or member is obsolete
+    }
+}
+
+public static class NatsTlsOptsExtensions
+{
+    public static SslClientAuthenticationOptions LoadClientCertFromPem(this SslClientAuthenticationOptions options, string certPem, string keyPem, bool offline = false, SslCertificateTrust? trust = null)
```

Comment:
@mtmk what is your opinion of putting these as extension methods on `SslClientAuthenticationOptions`?  Is that a good pattern or is there a better pattern to use here?

Reply:
agreed. would be much cleaner.